src/main/java/org/zeroturnaround/zip/ZipEntryUtil.java

```diff
@@ -78,20 +81,90 @@ static void copyEntry(ZipEntry zipEntry, InputStream in, ZipOutputStream out) th
    * Copies a given ZIP entry to a ZIP file. If this.preserveTimestamps is true, original timestamp
    * is carried over, otherwise uses current time.
    *
-   * @param zipEntry
+   * @param originalEntry
    *          a ZIP entry from existing ZIP file.
    * @param in
    *          contents of the ZIP entry.
    * @param out
    *          target ZIP stream.
    */
-  static void copyEntry(ZipEntry zipEntry, InputStream in, ZipOutputStream out, boolean preserveTimestamps) throws IOException {
-    ZipEntry copy = copy(zipEntry);
-    copy.setTime(preserveTimestamps ? zipEntry.getTime() : System.currentTimeMillis());
+  static void copyEntry(ZipEntry originalEntry, InputStream in, ZipOutputStream out, boolean preserveTimestamps) throws IOException {
+    ZipEntry copy = copy(originalEntry);
+
+    if (preserveTimestamps) {
+      /**
+       * If we succeed to set the modified time (works only on JDK8)
+       * then we shouldn't call the setTime method anymore as that
+       * will nullify the modified time - see
+       * http://hg.openjdk.java.net/jdk8u/jdk8u/jdk/file/f1f3f9eaf7fa/src/share/classes/java/util/zip/ZipEntry.java#l163
+       */
+      boolean success = tryToPreserveLatestJDKSupportedTimes(copy, originalEntry);
+      if (!success) {
+        copy.setTime(originalEntry.getTime());
+      }
+    }
+    else {
+      copy.setTime(System.currentTimeMillis());
+    }
     addEntry(copy, new BufferedInputStream(in), out);
   }
 
   /**
+   * The method tries to set the last modified time. The pre-requisite for this is that
```

Comment:
How can I get this not to use reflection over and over again? I can cache the class and method lookups but seems a very early optimisation. No?

It uses reflection over and over again only if you want to preserve timestamps. And if you want to preserve the timestamps and you do a lot of iterations JIT will save you again I believe.

As for your second comment that the code doesn't belong in the class I agree. The actual setting of timestamps and the logic belongs but all the underlying reflection doesn't.

I tried delegating method invocations to ZTZipReflectionUtil but at one point it felt just dumb.  What will I improve?